diagnostics/app/controllers/CommercialPreflightController.scala

```diff
@@ -0,0 +1,89 @@
+package controllers
+
+import conf.Configuration
+import common.{ExecutionContexts, Logging}
+import model.PrivateCache
+import org.joda.time.DateTime
+import play.api.libs.json.Json
+import play.api.libs.ws.WSClient
+import play.api.mvc.{Action, Controller}
+
+import java.lang.Long.{toString => toStringBase}
+import scala.util.{Success, Failure, Random}
+
+case class AdRequest(
+  load_id: String,
+  zone_ids: List[Int],
+  ip: String,
+  user_agent: String,
+  url: String,
+  referrer: Option[String],
+  browser_dimensions: Option[String],
+  switch_user_id: Option[String],       // switch_user_id, SWID cookie
+  floor_price: Option[String],
+  targeting_variables: Option[String]
+)
+
+object AdRequest {
+  implicit val adRequestWrites = Json.writes[AdRequest]
+}
+
+class CommercialPreflightController(wsClient: WSClient) extends Controller with Logging with ExecutionContexts {
+
+  // This is based on https://github.com/guardian/ophan/blob/master/tracker-js/assets/coffee/ophan/transmit.coffee
+  def makeOphanViewId(): String = {
+    val datePart = toStringBase(new DateTime().getMillis, 36)
+    val randPart = (for (_ <- 1 to 12) yield toStringBase(Random.nextInt(36), 36)).mkString
+    s"$datePart$randPart"
+  }
+
+  // This processes esi sub-requests from fastly.
+  def adCall() = Action { implicit request =>
+    val pageViewId = makeOphanViewId()
+
+    val maybeClientIp = request.headers.get("Fastly-Client-IP")
+    val maybeUserAgent = request.headers.get("User-Agent")
+    val maybeHost =  request.headers.get("Host")
+    val switchId = request.headers.get("X-GU-switch-id")
+    val maybeTopUrl = request.headers.get("X-GU-topurl")
+
+    for {
+      clientIp <- maybeClientIp
+      host <- maybeHost
+      topUrl <- maybeTopUrl
+      userAgent <- maybeUserAgent
+    } {
+      val adHubRequest = Json.toJson(AdRequest(
+        load_id = pageViewId,
+        zone_ids = List(229, 228),
+        /*  id    sizes
+            229   970x250, 728x90
+            228   300x250
```

Comment:
(And is there a list somewhere? Would here be a place to add the mappings for reference?)